cranelift/codegen/src/isa/unwind/winarm64.rs

```diff
@@ -0,0 +1,308 @@
+                    writer.write_u16_be(encoding);
+                }
+                &UnwindCode::LargeStackAlloc { size } => {
+                    // Size is measured in double 64-bit words.
+                    let mut encoding = size / 16;
+                    assert!(encoding < (1 << 24), "Stack alloc size too large");
+                    encoding |= 0b11100000 << 24;
+                    writer.write_u32_be(encoding);
+                }
+                UnwindCode::PacSignLr => {
+                    writer.write_u8(0b11111100);
+                }
+                UnwindCode::SetFp => {
+                    writer.write_u8(0b11100001);
+                }
+                &UnwindCode::AddFp { mut offset } => {
+                    offset /= 8;
+                    assert!(offset & !0xFF == 0, "Offset too large");
+                    let encoding = (0b11100010 << 8) | offset;
+                    writer.write_u16_be(encoding);
+                }
+            }
+        }
+    }
+}
+
+pub(crate) fn create_unwind_info_from_insts(
+    insts: &[(CodeOffset, UnwindInst)],
+) -> CodegenResult<UnwindInfo> {
+    let mut unwind_codes = vec![];
+    let mut last_stackalloc = None;
+    let mut last_clobber_offset = None;
+    for &(_, ref inst) in insts {
+        match inst {
+            &UnwindInst::PushFrameRegs { .. } => {
+                unwind_codes.push(UnwindCode::SaveFpLrPair { stack_offset: 16 });
+                unwind_codes.push(UnwindCode::SetFp);
+            }
+            &UnwindInst::DefineNewFrame {
+                offset_downward_to_clobbers,
+                ..
+            } => {
+                assert!(last_clobber_offset.is_none(), "More than one frame defined");
+                last_clobber_offset = Some(offset_downward_to_clobbers);
+
+                // If we've seen a stackalloc, then we were adjusting the stack
+                // to make space for additional arguments, so encode that now.
+                if let &Some(last_stackalloc) = &last_stackalloc {
+                    assert!(last_stackalloc < (1u32 << 8) * 8);
+                    unwind_codes.push(UnwindCode::AddFp {
+                        offset: last_stackalloc as u16,
+                    });
+                    unwind_codes.push(UnwindCode::SaveFpLrPair { stack_offset: 0 });
+                    unwind_codes.push(UnwindCode::SetFp);
+                }
+            }
+            &UnwindInst::StackAlloc { size } => {
+                last_stackalloc = Some(size);
+                assert!(size % 16 == 0, "Size must be a multiple of 16");
+                const SMALL_STACK_ALLOC_MAX: u32 = (1 << 5) * 16;
+                const MEDIUM_STACK_ALLOC_MAX: u32 = (1 << 11) * 16;
+                const LARGE_STACK_ALLOC_MAX: u32 = (1 << 24) * 16;
+                match size {
+                    0..SMALL_STACK_ALLOC_MAX => {
+                        unwind_codes.push(UnwindCode::SmallStackAlloc { size: size as u16 })
```

Comment:
At least in new code, we're trying to avoid `as`-casts; even though here it's infallibly correct, could we write `u16::try_from(size).unwrap()` (or `.expect("constant max size for this case exceeds u16")` or somesuch so that auditing for overflows is a little easier?

Reply:
Done: removes the `as` casts.